### PR TITLE
get_data() uses an external function to retrieve data from devices

### LINKED ISSUE #49: Allow an external procedure to do the read_from_device in the fields get_data()
This will allow to uncouple dl_esm_inf from the programming model used to communicate with the devices (currently OpenAcc and OpenCL), instead it will be the user of the library who needs to provide  and compile the function that the get_data method will use to retrieve the data.

This should make it easier to integrate other programming models like Kokkos and SYCL as the infrastructure library doesn't need to be aware of their specific implementation.

### PR DESCRIPTION
**Pull request by sergisiso**
Fixes #49

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This PR is ready for review. I created another PR in  PSycloneBench/57 to demonstrate its intended use, to look it you can use [its comparison with the latest kokkos PR](https://github.com/stfc/PSycloneBench/compare/kokkos_reduce_copies...read_device_function) as it is branched from it.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. I've learnt some new Fortran :-) 
I can build and run the example and the tests without problems.
Only thing that needs doing is documentation: the README needs the section on OpenCL deleting. Ideally we'd add some words to the actual documentation (under `doc`) instead. Maybe just a paragraph saying the field_type includes support for data in remote memory spaces and has the xxx functions as callbacks.

**Comment by sergisiso** [comment]
@arporter PR ready for next review.

> Only thing that needs doing is documentation: the README needs the section on OpenCL deleting. Ideally we'd add some words to the actual documentation (under doc) instead. Maybe just a paragraph saying the field_type includes support for data in remote memory spaces and has the xxx functions as callbacks.

Good point, I added a whole new section describing this functionality.

File: doc/api.rst [hunk 1]
```diff
@@ -134,6 +137,87 @@ call to ``grid_init`` for instance) before it is passed into this
...
+synchronization between the original data and the device data.
+
+These capabilities are provided by the following field attributes:
+
+ - `field_type%data_on_device`: A boolean to indicate if the data has already
+   been allocated and copied in the device.
+
+ - `field_type%read_from_device_f` or `field_type%read_from_device_c`: Function
+   pointers that provide the synchronization method to copy the data back from
+   the device into the host. The user needs to provide on one of these function
```
Comment (arporter) [inline doc/api.rst]: Should this say "...provide only one of..."? If so, I'll fix it and commit...

Comment (sergisiso) [reply doc/api.rst]: Ups, I wanted to say "...provide one of ..." but "only one" also works.

**Review by arporter — APPROVED** [review]
All looks good now - many thanks for adding the documentation.

**Comment by sergisiso** [comment]
@arporter Note that I made another commit to remove some OCL mentions in the Makefile after you approved the PR.

**Comment by arporter** [comment]
> @arporter Note that I made another commit to remove some OCL mentions in the Makefile after you approved the PR.

Thanks Sergi. I've just re-tested and grep'd for ocl and am happy. There's still a `set_data` subroutine with a TODO that refers to FortCL but I think we can leave that as is for the moment.

**Pull request merged**